feat: create recordWithPatterns schema

EskiMojo14 opened the pull request:
fixes #1159

```ts
const Schema = v.recordWithPatterns(
  [
    [
      v.pipe(v.string(), v.regex(/^foo\(.+\)$/)),
      v.pipe(v.string(), v.maxLength(1)),
    ],
    [
      v.pipe(v.string(), v.regex(/^bar\(.+\)$/)),
      v.pipe(v.string(), v.maxLength(10)),
    ],
  ],
  v.number(),
);
```

Also includes proper index signatures for strongly typed keys, e.g.
```ts
const FooKeySchema = v.custom<`foo(${string}`>(
  (arg) => typeof arg === "string" && /^foo\(.+\)$/.test(arg),
);

const Schema = v.recordWithPatterns([[FooKeySchema, v.string()]], v.number());

const parsed = v.parse(Schema, { "foo(x)": "x", x: 1 });

parsed["foo(x)"]; // string
parsed.x; // number
```

EskiMojo14 committed:
create objectWithPatterns schema

EskiMojo14 committed:
properly support transforming key

EskiMojo14 committed:
objectWithPatternsAsync

EskiMojo14 committed:
docs

fabian-hiller commented:
Thank you for creating this PR! Do you know if Zod offers such a schema? How about calling it `recordWithPatterns` or `templateRecord` to better match our `record` schema? Is the second argument, the wildcard schema, necessary? Maybe it simplifies the API if we ask users to add this to the first argument.

EskiMojo14 commented:
> Do you know if Zod offers such a schema? 

not to my knowledge, no - JSON schema has patternProperties which is similar

> How about calling it recordWithPatterns or templateRecord to better match our record schema?

sure, makes sense

> Is the second argument, the wildcard schema, necessary? Maybe it simplifies the API if we ask users to add this to the first argument.

what would be the desired behaviour if a key doesn't match any of the patterns?

EskiMojo14 committed:
objectWithPattern -> recordWithPattern

EskiMojo14 commented:
that does still leave the question of what to do when a key doesn't match any of the patterns:
```ts
const parsed = v.parse(v.recordWithPatterns([v.pipe(v.string(), v.startsWith("foo-")), v.string()]), { 'foo-allowed': 'hi', notAllowed: true })
// loose - { 'foo-allowed': 'hi', notAllowed: true }
// strip - { 'foo-allowed': 'hi' }
// strict - error
// rest - validate against rest, error if not match
```

EskiMojo14 committed:
address copilot comments

- check key validity before processing
- use untransformed value for issue reporting
- refer to "a record schema" instead of "an object schema" in comments and documentation
- add TODO note

EskiMojo14 commented on library/src/schemas/recordWithPatterns/recordWithPatternsAsync.ts:
```diff
@@ -0,0 +1,275 @@
+            dataset.typed = false;
+          }
+
+          // Add entry to dataset
+          // @ts-expect-error
+          dataset.value[newKey] = valueDataset.value;
```
this is a good point, but the current behaviour is consistent with `v.record` - both should be changed if this is a concern.

EskiMojo14 commented on library/src/schemas/recordWithPatterns/recordWithPatterns.ts:
```diff
@@ -0,0 +1,254 @@
+          let newKey = key;
+          for (const [keySchema, valueSchema_] of patterns) {
+            const keyDataset = keySchema['~run']({ value: key }, config);
+            if (keyDataset.typed) {
```
not a bad point, but current behaviour matches `v.record`

EskiMojo14 committed:
address AI feedback:

- add missing properties and type docs
- fix PatternTuple generics to be string instead of unknown
- remove unused expected type in RecordWithPatternsIssue

EskiMojo14 committed:
formatting
